File: service/src/main/java/com/boxfox/cross/service/friend/FriendService.java
package com.boxfox.cross.service.friend;

import com.boxfox.cross.service.JooqReactiveService;
import com.linkbit.android.entity.UserModel;
import io.reactivex.Completable;
import io.reactivex.Single;
import java.util.List;


public class FriendService extends JooqReactiveService implements FriendUsecase {

  private final FriendServiceImpl impl;

  public FriendService() {
    this.impl = new FriendServiceImpl();
  }

  @Override
  public Single<List<UserModel>> loadFriends(String uid) {
    return createSingle(ctx -> impl.loadFriends(ctx, uid));
  }

  @Override
  public Single<UserModel> getUser(String uid) {
    return createSingle(ctx -> impl.getUser(ctx, uid));
  }

  @Override
  public Completable addFriend(String ownUid, String uid) {
    return createCompletable(ctx -> impl.addFriend(ctx, ownUid, uid));
  }

  @Override
  public Completable deleteFriend(String ownUid, String uid) {
    return createCompletable(ctx -> impl.deleteFriend(ctx, ownUid, uid));
  }

  @Override
  public Single<List<UserModel>> serachUsers(String text) {
    return createSingle(ctx -> impl.searchUser(ctx, text));
  }
}

File: service/src/main/java/com/boxfox/cross/service/friend/FriendService.kt
package com.boxfox.cross.service.friend

import com.boxfox.cross.common.entity.UserModel
import com.boxfox.cross.service.JooqReactiveService
import io.reactivex.Completable
import io.reactivex.Single


class FriendService(private val impl: FriendServiceImpl = FriendServiceImpl()) : JooqReactiveService() {

    fun loadFriends(uid: String): Single<List<UserModel>> {
        return single { ctx -> impl.loadFriends(ctx, uid) }
    }

    fun getUser(uid: String): Single<UserModel> {
        return single { ctx -> impl.getUser(ctx, uid) }
    }

    fun addFriend(ownUid: String, uid: String): Completable {
        return completable { ctx -> impl.addFriend(ctx, ownUid, uid) }
    }

    fun deleteFriend(ownUid: String, uid: String): Completable {
        return completable { ctx -> impl.deleteFriend(ctx, ownUid, uid) }
    }

    fun serachUsers(text: String): Single<List<UserModel>> {
        return single { ctx -> impl.searchUser(ctx, text) }
    }
}
